<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению
<center>Автор материала: Трунов Артем Геннадьевич, @datamove.

# <center>Pipeline, FeatureUnion – практика применения

## <center>Введение</center>
В этой статье будем разбираться с классами пакета sklearn, которые представляют значительное удобство и экономию времени в работе. Многие, наверное, любят, когда код иллюстрируется диаграммами классов или каким-нибудь метакодом, который позволяет убрать из поля зрения все детали реализации и оставить на виду только самое главное. Pipeline в sklearn - это и есть такой вот метакод, с помошью которого модель видна как на ладони.

В качестве же примеров будем использовать не измусоленные со всех сторон встроенные в sklearn датасеты, а знакомые читателю по домашним работам и соревнованию 'Alice' данные. Надеюсь, что кого-то эта статья побудит исправить свой код и вдохновит на новые засылки на Kaggle!

## <center>Pipeline</center>
Итак, начнем с перевода и определения. Русские слова "труба" и,тем паче, "трубопровод", мы, пожалуй, использовать не будем, а вот вариант "конвейер данных" кажется мне наиболее подходящим и благозвучным.

Документация Pipeline определяет этот класс как конвейер преобразования данных с финальным эстиматором (обучающей моделью), применяющийся для того, чтобы можно было легко менять параметры на каждом этапе конвейера и сравнивать результаты. С такой же легкостью можно заменять и сами этапы преобразований данных и финальную модель.

Давайте сразу окунемся в пример. Рассмотрим датасет Самсунга из домашней работы №7. Мы применяли  к данным алгоритм PCA для уменьшения размерности, а что бы он работал как надо, предварительно масштабировали данные. Для классификации использовали метод опорных векторов. Таким образом, наш конвейер будет состоять из двух шагов обработки (StandardScaler, PCA) и финальной модели (LinearSVC).

In [1]:
#изменить соответственно
PATH_TO_DATA="../../"

In [2]:
#загрузка данных
#На всякий случай - ссылка https://cloud.mail.ru/public/3EJK/cB2VXsyrP
import numpy as np

X_train = np.loadtxt(PATH_TO_DATA+"data/samsung_HAR/samsung_train.txt")
y_train = np.loadtxt(PATH_TO_DATA+"data/samsung_HAR/samsung_train_labels.txt").astype(int)

X_test = np.loadtxt(PATH_TO_DATA+"data/samsung_HAR/samsung_test.txt")
y_test = np.loadtxt(PATH_TO_DATA+"data/samsung_HAR/samsung_test_labels.txt").astype(int)

In [3]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.svm import LinearSVC

pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA(n_components=65)),
    ('svc', LinearSVC())
])

Ну как, красиво? Давайте разберем. Конструктор Pipeline() принимает массив кортежей, в каждом из которых мнемоническое обозначение этапа преобразования и экземпляр класса преобразователя, инстанциированный "на лету". Первый этап предразования, 'scaler', принимает исходные данные, и выдает отмасштабированные на выход, который является входом второго этапа - 'pca'. Из 'pca' в 'svc' поступает урезанная матрица главных компонентов числом 65 штук. Обучение проводится именно на ней.

Давайте запустим обучение и получим результат.


In [4]:
pipeline.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scaler', StandardScaler(copy=True, with_mean=True, with_std=True)), ('pca', PCA(copy=True, iterated_power='auto', n_components=65, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('svc', LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))])

In [5]:
#"Валидируемся" на том же тренировочном датасете
pred = pipeline.predict(X_train)

from sklearn.metrics import roc_auc_score, accuracy_score

accuracy_score(pred, y_train)

0.95661044613710555

Заметим, что метод fit() вызывается на всех этапах конвейера, а метод predict() - только для финального эстиматора. Тоже самое, разумеется, происходит и с другим набором данных:

In [6]:
#Валидируемся на тестовом датасете, для которого у нас есть разметка
test_pred = pipeline.predict(X_test)
accuracy_score(test_pred, y_test)

0.92263318629114355

Это значит, что нам не надо тащить за собой хвост из преобразований тестовой выборки, об этом позаботится наш конвейер!

Так как конвейер обладает интерфейсом модели обучения (fit(), predict() etc), то мы можем использовать его напрямую с полюбившимися методами, такими как кросс-валидация:

In [7]:
from sklearn.model_selection import cross_val_score

cross_val_score(pipeline, X_train, y_train, cv=3)

array([ 0.91761827,  0.86460033,  0.91993464])

а также GridSearchCV:

In [8]:
from sklearn.model_selection import GridSearchCV

gcv_params = {'pca__n_components': [20,60,100], 
              'svc__C': [0.001, 0.01, 0.1, 1, 10] }

gcv = GridSearchCV(pipeline, gcv_params, cv=3)
gcv.fit(X_train,y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', StandardScaler(copy=True, with_mean=True, with_std=True)), ('pca', PCA(copy=True, iterated_power='auto', n_components=65, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('svc', LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'pca__n_components': [20, 60, 100], 'svc__C': [0.001, 0.01, 0.1, 1, 10]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [9]:
gcv.best_params_

{'pca__n_components': 100, 'svc__C': 0.1}

"Так," - скажет внимательный читатель,- "а ведь нам надо было выбрать число компонент так, чтобы оставить 90% дисперсии исходных данных". Как же мы используем это условие в конвейере?". Вообще-то оно уже реализовано в классе PCA - достаточно передать параметер n_components=0.9 в конструктор класса. Но давайте сделаем сами. Для этого нам придется реализовать собственный класс эстиматора, для многих - первый в их жизни! Сейчас увидим, что на самом деле, это - легко!

Мы унаследуем класс PCA и перегрузим методы fit(), transform() и fit_tranaform() так, чтобы возвращать матрицу с числом компонентов, объясняющих exp_var% дисперсии.

Затем построим конвейер с новым классом.

In [10]:
class PCAExplainedVariance(PCA):
    #констуктор принимает и сохраняет значение желаемой дисперсии
    def __init__(self, exp_var=1.0 ):
        super().__init__(copy=True)
        self.exp_var = exp_var #желаемая дисперсия исходных данных
        self.N_ = 0 #число компонент, тербуемых для достижения заданной дисперсии

    # Находим соответствующее число компонент
    def fit(self, X, y=None):
        super().fit(X, y)
        self.N_ = len(X)
        cum_var = 0
        for i, component in enumerate(self.components_):
            cum_var += self.explained_variance_ratio_[i]
            if cum_var>=self.exp_var:
                self.N_ = i + 1
                break
              
    # возвращаем усеченный по числу компонент датасет
    def transform(self, X, y=None):
        U = X[:,:self.N_]
        return U
    
    # fit + transform в одном флаконе
    def fit_transform(self, X, y=None):
        self.fit(X)
        U = X[:, :self.N_]

        return U

In [11]:
#Снова собираем конвейер
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCAExplainedVariance(exp_var=0.9)),
    ('svc', LinearSVC())
])

In [12]:
#На этот раз запустим с GridSearchCV
gcv_params = {'svc__C': [0.001, 0.01, 0.1, 1, 10] }
gcv = GridSearchCV(pipeline, gcv_params, cv=3)
gcv.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('scaler', StandardScaler(copy=True, with_mean=True, with_std=True)), ('pca', PCAExplainedVariance(exp_var=0.9)), ('svc', LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'svc__C': [0.001, 0.01, 0.1, 1, 10]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [13]:
gcv.best_params_

{'svc__C': 0.1}

Объект конвейера предоставляет доступ и к экземплярам составляющих его классов. Например, чтобы посмотреть, какое число компонент оставил наш новый PCA-эстиматор:

In [14]:
gcv.best_estimator_.named_steps['pca'].N_

63

## <center>Feature Union</center>

Давайте идти дальше и расширять диапазон применяемых средств. Для этого возьмем в качестве примера более сложный случай.

В соревновании <a href="https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2">Catch me if you can</a> (aka "Alice") на Kaggle, мы отдельно обрабатываем посещаемые пользователями сайты с помощью техники Bag of Words, и отдельно конструируем новые признаки из чего только можно. Затем объединяем частотную матрицу с матрицей признаков и применяем логистическую регрессию.

Попробуем запрограммировать этот сценарий в конвейер.


In [15]:
#Загрузка и предобработка данных - код от @yorko
import pandas as pd
train_df = pd.read_csv(PATH_TO_DATA+"../Alice-comp/train_sessions.csv", index_col="session_id")
#test_df = pd.read_csv(PATH_TO_DATA+"../Alice-comp/test_sessions.csv", index_col="session_id")
        
# приведем колонки time1, ..., time10 к временному формату
times = ['time%s' % i for i in range(1, 11)]
train_df[times] = train_df[times].apply(pd.to_datetime).fillna(method='ffill', axis=1)
#test_df[times] = test_df[times].apply(pd.to_datetime).fillna(method='ffill', axis=1)

# отсортируем данные по времени
train_df = train_df.sort_values(by='time1')

sites = ['site%s' % i for i in range(1,11)]
train_df[sites] = train_df[sites].fillna(0).astype('int')
#test_df[sites] = test_df[sites].fillna(0).astype('int')

#целевая переменая
y_train = train_df['target']
train_df.drop('target', axis=1, inplace=True)

train_df.head()

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57     55 2013-01-12 08:05:57      0   
54843          56 2013-01-12 08:37:23     55 2013-01-12 08:37:23     56   
77292         946 2013-01-12 08:50:13    946 2013-01-12 08:50:14    951   
114021        945 2013-01-12 08:50:17    948 2013-01-12 08:50:17    949   
146670        947 2013-01-12 08:50:20    950 2013-01-12 08:50:20    948   

                         time3  site4               time4  site5  \
session_id                                                         
21669      2013-01-12 08:05:57      0 2013-01-12 08:05:57      0   
54843      2013-01-12 09:07:07     55 2013-01-12 09:07:09      0   
77292      2013-01-12 08:50:15    946 2013-01-12 08:50:15    946   
114021     2013-01-12 08:50:18    948 2013-01-12 08:50:18    945   
146670     2013-01-12 08:50:20    947 2013-01-12 08:50:21    950   

                         time5  site6               time6  site7  \
session_id                                                         
21669      2013-01-12 08:05:57      0 2013-01-12 08:05:57      0   
54843      2013-01-12 09:07:09      0 2013-01-12 09:07:09      0   
77292      2013-01-12 08:50:16    945 2013-01-12 08:50:16    948   
114021     2013-01-12 08:50:18    946 2013-01-12 08:50:18    947   
146670     2013-01-12 08:50:21    952 2013-01-12 08:50:21    946   

                         time7  site8               time8  site9  \
session_id                                                         
21669      2013-01-12 08:05:57      0 2013-01-12 08:05:57      0   
54843      2013-01-12 09:07:09      0 2013-01-12 09:07:09      0   
77292      2013-01-12 08:50:16    784 2013-01-12 08:50:16    949   
114021     2013-01-12 08:50:19    945 2013-01-12 08:50:19    946   
146670     2013-01-12 08:50:21    951 2013-01-12 08:50:22    946   

                         time9  site10              time10  
session_id                                                  
21669      2013-01-12 08:05:57       0 2013-01-12 08:05:57  
54843      2013-01-12 09:07:09       0 2013-01-12 09:07:09  
77292      2013-01-12 08:50:17     946 2013-01-12 08:50:17  
114021     2013-01-12 08:50:19     946 2013-01-12 08:50:20  
146670     2013-01-12 08:50:22     947 2013-01-12 08:50:22

Итак, у нас есть такой вот датафрейм и мы хотим:

    а) составить Bag Of Words из сайтов - код взят из ноутбука @yorko
    б) нагенерить признаки, связанные со временем, любезно подсказанные @yorko: year_month, start_hour, morning (последний признак - бинарный)
    
Реализуем a), б) по отдельности как классы-трансформеры, а потом объединим результаты.


In [16]:
# Этот класс-трансформер возвращает разреженную матрицу сайтов
#
from sklearn.base import BaseEstimator, TransformerMixin
from scipy.sparse import csr_matrix
class ColsToCountMatrix(BaseEstimator, TransformerMixin):
    #констуктор принимает и сохраняет название колонок для сливания в текст
    def __init__(self, columns=[]):
        self.columns=columns
        
    # fit() ничего не делает
    def fit(self, X, y = None):
        return self
    
    #преобразуем посещения сайтов в частотную матрицу
    def transform(self, X):
        # последовательность с индексами
        sites_flatten = X[self.columns].values.flatten()

        # искомая матрица
        sites_sparse = csr_matrix(([1] * sites_flatten.shape[0],
                                    sites_flatten,
                                    range(0, sites_flatten.shape[0] + 10, 10)))[:, 1:]
        return sites_sparse
#Unit test
sparse_matrix = ColsToCountMatrix(columns=sites).transform(train_df.head(3))
print(sparse_matrix.shape)
print(sparse_matrix)

(3, 951)
  (0, 55)	1
  (0, 54)	1
  (1, 55)	1
  (1, 54)	1
  (1, 55)	1
  (1, 54)	1
  (2, 945)	1
  (2, 945)	1
  (2, 950)	1
  (2, 945)	1
  (2, 945)	1
  (2, 944)	1
  (2, 947)	1
  (2, 783)	1
  (2, 948)	1
  (2, 945)	1


In [17]:
# Этот класс-трансформер возвращает матрицу с новыми признаками
#
from sklearn.base import BaseEstimator, TransformerMixin
class TimeToFeatures(BaseEstimator, TransformerMixin):
    # берем и сохраняем колонки, которые используем для приготовления новых признаков
    def __init__(self, columns=[]):
        self.columns = columns
    # бездельник опять
    def fit(self,X,y=None):
        return self
    # работяга
    def transform(self, X):
        # это колонка 'time1' начального датафрейма
        time1=self.columns[0] 
        # создаем пустой датафрейм для новых признаков
        new_features = pd.DataFrame(index=X.index)
        # делаем новые признаки
        new_features['year_month'] = X[time1].apply(lambda ts: ts.year*100 + ts.month)
        new_features['start_hour'] = X[time1].apply(lambda ts: ts.hour)
        new_features['morning'] = new_features['start_hour'].apply(lambda sh: 1 if 4<sh<12 else 0)
        return new_features[['year_month','start_hour','morning']]
#Unit test
TimeToFeatures(columns=times).transform(train_df.head()).values

array([[201301,      8,      1],
       [201301,      8,      1],
       [201301,      8,      1],
       [201301,      8,      1],
       [201301,      8,      1]])

Давайте теперь применим FeatureUnion. Конструктор класса FeatureUnion, как и конструктор Pipeline, принимает список кортежей (название, класс-трансформер), а его метод transform() просто объединяет колонки, получившиеся после применения метода transform() для каждого из составных классов.


In [18]:
from sklearn.pipeline import FeatureUnion

fu=FeatureUnion([
            ('cols_to_text', ColsToCountMatrix(columns=sites)),
            ('time_to_features', TimeToFeatures(columns=times)),
])

In [19]:
#используем todense() для наглядности
full_matrix = fu.transform(train_df.head(3)).todense()
print(full_matrix.shape)
print(full_matrix)

(3, 954)
[[     0      0      0 ..., 201301      8      1]
 [     0      0      0 ..., 201301      8      1]
 [     0      0      0 ..., 201301      8      1]]


В итоге - было 951 колонка частотной матрицы, 3 колонки новых признаков, стало 954 колонки.

Это еще не все! Раскроем возможность использовать FeatureUnion и Pipeline вместе.
Сразу поразим читателя, добавив этапы преобразования полученных данных, а так же модель обучения на объединенных данных.


In [20]:
from sklearn.pipeline import FeatureUnion
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegressionCV

logit_params={'scoring':'roc_auc','class_weight':'balanced',
             'Cs':range(1,5),'n_jobs':3, 'random_state':17}

#используем немного другой формат вызова FeatureUnion, 
#хотя веса для нашей модели не пригодятся, читатель будет знать о таких возможностях
pipeline = Pipeline([
    ('union', FeatureUnion(
        transformer_list=[
            ('text', Pipeline([
                ('cols_to_text', ColsToCountMatrix(columns=sites)),
                ('tfidf',TfidfTransformer()), 
            ])),
            ('new_features', Pipeline([
                ('time_to_features', TimeToFeatures(columns=times)),
                ('scaler', StandardScaler()),
            ])),
        ],
        transformer_weights={'text':1.0, 'features':1.0}
    )),
    ('logit',LogisticRegressionCV(**logit_params))
])

In [21]:
pipeline.fit(train_df, y_train)

Pipeline(memory=None,
     steps=[('union', FeatureUnion(n_jobs=1,
       transformer_list=[('text', Pipeline(memory=None,
     steps=[('cols_to_text', ColsToCountMatrix(columns=['site1', 'site2', 'site3', 'site4', 'site5', 'site6', 'site7', 'site8', 'site9', 'site10'])), ('tfidf', TfidfTransformer(norm='l2', smooth_idf=True,... random_state=17, refit=True, scoring='roc_auc', solver='lbfgs',
           tol=0.0001, verbose=0))])

In [22]:
# это подобранный перебором коэффициент регуляризации
pipeline.named_steps['logit'].C_

array([1])

In [23]:
#таблица метрики ROC_AUC для С=[1,2,3,4] и трех выборок кросс-валидации (cv=3)
pipeline.named_steps['logit'].scores_

{1: array([[ 0.90993035,  0.91043374,  0.91055791,  0.91065063],
        [ 0.91248111,  0.91011061,  0.90696356,  0.90567206],
        [ 0.93503879,  0.93541084,  0.93466052,  0.93417007]])}

## <center>Заметки</center>

Что ж, неплохо получилось! Наша модель описывается 16-ю строками, после того как мы реализовали преобразования данных в классах-трансформерах. 

Давайте теперь разберем некоторые вопросы применения конвейеров и объединителей признаков.

1. Для того, чтобы сделать предсказания обученной модели для тестовой выборки, вызовите метод pipeline.predict_proba(df_proba)

2. Мы не можем (по крайней мере, с легкостью) в нашем конвейере сделать чаcтотную матрицу на объединенной тренировочной и тестовой выборках, как @yorko делал это на мастер-классе. Автор решил эту проблему таким образом. Вместо класса ColsToCountMatrix, который работает с колонками sites, используем класс ColsToText, определенный ниже. Он собирает сайты из всех колонок в "текст", который принимает библиотечный CountVectorizer. В констукторе этого класса читатель найдет не только опцию vocabulary для передачи словаря объединенной тренировочной и тестовой выборки, но и некоторые опции, которые имеет смысл попробовать для улучшения результатов модели.

3. Если читатель решит применить другую модель обучения, например SGDClassifier, в котором не реализована кросс-валидация, то можно "обернуть" его в GridSearchCV:

    <p>('gcv', GridSearchCV(SGDClassifier(**sgd_params), gcv_sgd_params, **gcv_params))</p>
    

Надеюсь, что читатель сможет теперь сам улучшать свою модель для соревнования - работать над признаками и подбирать параметры.


## <center>В заключение</center>
Что можно посоветовать читателю в плане дальнейшего изучения предмета?

1. Изучить официальную документацию: <a href="http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html">PipeLine</a>, <a href="http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.FeatureUnion.html">FeatureUnion</a>, и разобрать статьи-примеры.
  
2. Взять на вооружение библиотеку <a href="https://github.com/rasbt/mlxtend">mlextend</a> Себастьяна Рашки. Там можно найти много интересных классов, не реализованных в стандартной библиотеке sklearn.

3. Посмотреть <a href="https://github.com/scikit-learn-contrib/sklearn-pandas">sklearn-pandas</a> - облегчение работы именно с датафреймами. Например, можно некоторые колонки преобразовать масштабированием, другие - по принципу one-hot-encoding.

4. Стремиться создавать такие конвейеры, которые позволяют быстро проверять модели и признаки.

Успехов!

In [24]:
#это подготовительный этап трансформации данных, 
#перед тем как применим CountVectorizer
from sklearn.base import BaseEstimator, TransformerMixin

class ColsToText(BaseEstimator, TransformerMixin):
    #конструктор принимает и сохраняет название колонок для сливания в текст
    def __init__(self,columns=[]):
        self.columns = columns
    # fit() отдыхает - делать нечего
    def fit(self, X, y= None):
        return self
    # сливаем содержимое колонок в одну строку, кроме нулей
    def transform(self, X):
        return X[self.columns]\
               .apply(lambda x: " ".join([str(a) for a in x.values if not a==0]), axis=1)\
               .values.reshape(len(X),1)
    #заметьте - возвращаем numpy.ndarray
# Unit test
ColsToText(columns=sites).transform(train_df.head())

array([['56 55'],
       ['56 55 56 55'],
       ['946 946 951 946 946 945 948 784 949 946'],
       ['945 948 949 948 945 946 947 945 946 946'],
       ['947 950 948 947 950 952 946 951 946 947']], dtype=object)